### DISCUSSION TITLE
New Nerd Font UTF-32 ranges + YAML = errors with UTF-16 surrogate pairs

### DISCUSSION THREAD

**Discussion by Destroy666x** [JanDeDobbeleer/oh-my-posh#3470] [General]
Related issue: https://github.com/go-yaml/yaml/issues/279

Probably worth noting that somewhere, so that people don't go for YAML, as that doesn't seem like it'll be fixed anytime soon.

Also, I checked the `oh-my-posh config migrate glyphs --write` command on my YAML config and it output a .yaml file in JSON format, which I'm not sure is something that can be expected.

**Comment by JanDeDobbeleer** [comment]
@Destroy666x migration is fixed

**Comment by Destroy666x** [comment]
@JanDeDobbeleer thanks for looking into it, but not too sure it is. Well, the extension is .yml now, however:
![image](https://user-images.githubusercontent.com/3606072/217722642-5335d793-fa9f-47b0-91f5-71b382a78385.png)

It seems to use single quotes which don't work with Unicode escape sequences. And the problem from the title and Go YAML lib is not really solved if you change to double quotes.

**Reply by Destroy666x** [reply]
And that repo seems to be pretty dead as well. Last PR merge is in Jan 2021.

**Reply by JanDeDobbeleer** [reply]
@Destroy666x it seems that this isn't supported for yaml, we've been here before and it just doesn't work regardless of the parser (just tried). They will need to fix that. My comment was oriented towards the JSON format in your YAML.

**Reply by Destroy666x** [reply]
Within that library itself, yeah. Not sure about any other for Go, but e.g. JS one supports them according to specification.

**Reply by Destroy666x** [reply]
As I also mentioned, I think you should document that YAML is not a recommended format due to this restriction.

**Reply by JanDeDobbeleer** [reply]
@Destroy666x or, I can fix upstream.

**Reply by Destroy666x** [reply]
I'm not sure I follow. Upstream like by forking the YAML library until they fix it? If so, that would be awesome indeed, since they don't push out any changes anyways.

**Reply by JanDeDobbeleer** [reply]
@Destroy666x yup, that's what I mean. Also found a more up-to-date implementation, and I have a fix ready on that one as well. Guess I'll have some PR's to do. I'll make sure to do the changes here first with forks so we can iron out the quirks and then see if we can get that merged upstream.

**Reply by JanDeDobbeleer** [reply]
@Destroy666x 
- https://github.com/gookit/config/pull/119
- https://github.com/goccy/go-yaml/pull/343
- https://github.com/JanDeDobbeleer/oh-my-posh/pull/3481

**Reply by Destroy666x** [reply]
Thanks. Now it's closer to working. One problem that's there for sure is that the surrogate pairs don't get quoted, so you still get a config error for unquoted `\`. So e.g. `"\uf17a"` got quoted correctly now, but `\udb81\ude3f` didn't. After quoting it manually the config doesn't throw an error though, unlike with the other library. I'll test if the characters work well by updating my fonts now, but it seems like they aren't parsed into a single character as 2 squares are shown - not sure if that's how it's supposed to work if the non-BMP character is unavailable in font regardless.

**Reply by Destroy666x** [reply]
No, there doesn't seem to be any output for the surrogate pairs if you use latest nerd fonts, I tried with a few, just 2 squares still.

**Reply by JanDeDobbeleer** [reply]
@Destroy666x I'll have a look as I also noticed that. In JSON this is part of the spec, don't know for YAML.

**Reply by Destroy666x** [reply]
YAML has to quote `\` always I believe according to spec, but not 100% sure if it's a requirement or just a recommendation. Either with `''` in regular string or with `""` if escape sequences should be interpreted as such. But regardless, escape sequences need to be 100% double quoted.

**Reply by Destroy666x** [reply]
I assume still no complete solution found @JanDeDobbeleer ? And does it work with the new Nerd Font Material codepoints and JSON? If so, in which shell? Can any of Windows shells display UTF-32 in the 1st place? I made a PowerShell script to try to display alll the NF characters and that range doesn't work in cmd at all as expected, but also in Powershell I can't get it to display, even if you set e.g. `$OutputEncoding` to `[System.Text.Encoding]::UTF32`

**Reply by JanDeDobbeleer** [reply]
@Destroy666x works in JSON, I fixed part of this, can't be held responsible to fix golang's YAML libs. I want to have some sort of life 😅

**Reply by Destroy666x** [reply]
I see. I'll just go back to old NFs then until some fix appears some day, the new ones seem quite problematic.
